Assets/Scripts/RedRunner/Characters/RedCharacter.cs

```diff
@@ -507,19 +529,41 @@ public override void Move ( float horizontalAxis )
 
 		public override void Jump ()
 		{
-			if ( !IsDead.Value && m_JumpsSoFar < m_AllowedJumps)
+			if (!IsDead.Value)
 			{
-				m_JumpsSoFar++;
-				Vector2 velocity = m_Rigidbody2D.velocity;
-				velocity.y = m_JumpStrength;
-				m_Rigidbody2D.velocity = velocity;
-				m_Animator.animator.ResetTrigger ( "Jump" );
-				m_Animator.SetTrigger ( "Jump" );
-				m_JumpParticleSystem.Play ();
-				AudioManager.Singleton.PlayJumpSound ( m_JumpAndGroundedAudioSource );
+				if (m_GroundCheck.IsGrounded || m_WallDetector.TouchingWall)
+				{
+					ApplyJumpPhysics(m_JumpStrength);
+				}
+				else if (m_HasDoubleJump)
+				{
+					ApplyJumpPhysics(m_DoubleJumpStrength);
+				}
 			}
 		}
 
+		private void ApplyJumpPhysics(float jumpStrength)
+		{
+			Vector2 velocity = m_Rigidbody2D.velocity;
+			velocity.y = jumpStrength;
+			m_Rigidbody2D.velocity = velocity;
+			m_Animator.animator.ResetTrigger("Jump");
+			m_Animator.SetTrigger("Jump");
+			m_JumpParticleSystem.Play();
+			AudioManager.Singleton.PlayJumpSound(m_JumpAndGroundedAudioSource);
+			m_HasDoubleJump = false;
```

Comment:
m_HasDoubleJump = false; should prob stay outside the helper (only in the else case)

Reply:
whoops, good catch